File: Project1/src/project1/Zookeeper.java
package project1;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import project1.animals.Animal;

public class Zookeeper {
  /**
   *  Default constructor
   */
  public Zookeeper(){}

  /**
   *  @param zoo Zoo
   *  Iterate through zoo animals and wakes each one
   */
  public void wakeAnimals(Zoo zoo){
    displayAction("Zookeeper is waking the animals...");
    List<Animal> animals = zoo.getAnimals();
    for (Animal a : animals) {
			a.wakeUp();
		}
  }

  /**
   *  @param zoo Zoo
   *  Iterate through zoo animals and calls roll for each one
   */
  public void rollCall(Zoo zoo){
    displayAction("Zookeeper is calling roll...");
    List<Animal> animals = zoo.getAnimals();
    for (Animal a : animals) {
			a.makeNoise();
		}
  }

  /**
   *  @param zoo Zoo
   *  Iterate through zoo animals and feeds each one
   */
  public void feedAnimals(Zoo zoo){
    displayAction("Zookeeper is feeding the animals...");
    List<Animal> animals = zoo.getAnimals();
    for (Animal a : animals) {
			a.eat();
		}
  }

  /**
   *  @param zoo Zoo
   *  Iterate through zoo animals and exercises each one
   */
  public void exerciseAnimals(Zoo zoo){
    displayAction("Zookeeper is exercising the animals...");
    List<Animal> animals = zoo.getAnimals();
    for (Animal a : animals) {
			a.roam();
		}
  }

  /**
   *  @param zoo Zoo
   *  Iterate through zoo animals and puts each one to bed
   */
  public void closeZoo(Zoo zoo){
    displayAction("Zookeeper is closing the zoo and puting the animals to bed...");
    List<Animal> animals = zoo.getAnimals();
    for (Animal a : animals) {
			a.sleep();
		}
  }

  /**
   *  @param action String
   *  Helper function to show start of new action
   */
  private void displayAction(String action){
    System.out.println("* * * * * *");
    System.out.println(action);
    System.out.println("* * * * * *");
  }

}

File: Project2/src/Zookeeper.ts
import { Animal } from "./animals";
import { Zoo } from "./Zoo";
import Subject from "./Subject"
import Observer from "./Observer"

export enum ZookeeperActionState {
  WakingAnimals,
  TakingRoll,
  FeedingAnimals,
  ExersizingAnimals,
  ClosingZoo
}

export class Zookeeper implements Subject {

  // Define observer variables
  private state: ZookeeperActionState;

  getZookeeperActionState(): ZookeeperActionState {
    return this.state;
  }
  setZookeeperActionState(newZookeeperActionState: ZookeeperActionState) {
    this.state = newZookeeperActionState;
    this.notifyObserver();
  }
  observer: Observer = null;
  notifyObserver() {
    if(this.observer != null)
      this.observer.update();
  }
  setObserver(newObserver: Observer) {
    this.observer = newObserver;
  }
  removeObserver() {
    this.observer = null;
  };

  private zoo: Zoo;


  /**
   *  Default constructor
   */
  constructor(zoo: Zoo) {
    this.zoo = zoo;
  }

  /**
   *  @param zoo Zoo
   *  Iterate through zoo animals and wakes each one
   */
  public wakeAnimals() {
    this.setZookeeperActionState(ZookeeperActionState.WakingAnimals);
    this.displayAction("Zookeeper is waking the animals...");
    var animals = this.zoo.getAnimals();
    animals.forEach(function(animal) {
      animal.wakeUp();
    });
  }

  /**
   *  @param zoo Zoo
   *  Iterate through zoo animals and calls roll for each one
   */
  public rollCall() {
    this.setZookeeperActionState(ZookeeperActionState.TakingRoll);
    this.displayAction("Zookeeper is calling roll...");
    var animals: Array<Animal> = this.zoo.getAnimals();
    animals.forEach(function(animal) {
      animal.makeNoise();
    });
  }

  /**
   *  @param zoo Zoo
   *  Iterate through zoo animals and feeds each one
   */
  public feedAnimals() {
    this.setZookeeperActionState(ZookeeperActionState.FeedingAnimals);
    this.displayAction("Zookeeper is feeding the animals...");
    var animals: Array<Animal> = this.zoo.getAnimals();
    animals.forEach(function(animal) {
      animal.eat();
    });
  }

  /**
   *  @param zoo Zoo
   *  Iterate through zoo animals and exercises each one
   */
  public exerciseAnimals() {
    this.setZookeeperActionState(ZookeeperActionState.ExersizingAnimals);
    this.displayAction("Zookeeper is exercising the animals...");
    var animals: Array<Animal> = this.zoo.getAnimals();
    animals.forEach(function(animal) {
      animal.roam();
    });
  }

  /**
   *  @param zoo Zoo
   *  Iterate through zoo animals and puts each one to bed
   */
  public closeZoo() {
    this.setZookeeperActionState(ZookeeperActionState.ClosingZoo);
    this.displayAction("Zookeeper is closing the zoo and puting the animals to bed...");
    var animals: Array<Animal> = this.zoo.getAnimals();
    animals.forEach(function(animal) {
      animal.sleep();
    });
    this.removeObserver();
  }

  /**
   *  @param action String
   *  Helper function to show start of new action
   */
  private displayAction(action: String) {
    console.log("* * * * * *");
    console.log(action);
    console.log("* * * * * *");
  }

}
